token/src/token.rs

```diff
@@ -0,0 +1,190 @@
+use bulletproofs::{BulletproofGens, PedersenGens};
+use curve25519_dalek::ristretto::CompressedRistretto;
+use curve25519_dalek::scalar::Scalar;
+use zkvm::{
+    Commitment, Data, Entry, Input, Instruction, Predicate, Program, Prover, Signature, Tx,
+    TxHeader, TxID, TxLog, VMError, Value, VerificationKey, Verifier,
+};
+
+/// Represents a ZkVM Token with unique flavor and embedded
+/// metadata protected by a user-supplied Predicate.
+pub struct Token {
+    issuance_predicate: Predicate,
+    metadata: Vec<u8>,
+}
+
+impl Token {
+    /// Constructs a new Token.
+    pub fn new(pred: Predicate, metadata: Vec<u8>) -> Self {
+        Token {
+            issuance_predicate: pred,
+            metadata: metadata,
+        }
+    }
+
+    /// Returns program that issues specified quantity of Token.
+    pub fn issue<'a>(program: &'a mut Program, token: &Token, qty: u64) -> &'a mut Program {
+        program
+            .push(Commitment::blinded(qty)) // stack: qty
+            .var() // stack: qty-var
+            .push(Commitment::unblinded(Value::issue_flavor(
```

Comment:
How about factoring flavor computation in a `pub fn flavor(&self) -> Scalar` so that the user of `Token` can always ask for its flavor?